LoadLibrary for Lazy P/Invoke Improvements

shrah opened the pull request:
This change improves the loadlibrary for Lazy P/Invoke by propagating
DllImportSearchPath attribute value to runtime and tries to imitate the
CoreCLR behavior. In particular, this change ensures we look for native
libraries in the directory where the executable resides by default.

jkotas commented on src/Native/System.Private.CoreLib.Native/pal_exepath.cpp:
```diff
@@ -0,0 +1,93 @@
+#define symlinkEntrypointExecutable "/proc/curproc/exe"
+#endif
+
+extern "C" bool CoreLibNative_GetEntrypointExecutableAbsolutePath(char **buf)
```
Is this what we are doing in CoreCLR? I thought that we do the custom load library handling on Windows only because of that problem really only exist on Windows.

Also, piggy backing on the .exe name is not compatible with standalone .dlls that we want to support in CoreRT eventually.

jkotas replied:
I do not see any code like this under https://github.com/dotnet/coreclr/tree/master/src/pal .

sergiy-k replied:
dlopen not does include the current app directory in library search path. In CoreCLR this (and platform specific prefix and extenstions) are handled here: https://github.com/dotnet/coreclr/blob/master/src/vm/dllimport.cpp#L6223

jkotas replied:
This behavior makes sense once we have optional loading of IL ECMA assemblies in CoreRT. 

I am wondering the simulator of the probing path makes sense at all for the native binaries produced by the full AOT. Why can't we just set RPATH on the native binary? It is what how one would deal with this problem on regular native binaries on Unix.

sergiy-k replied:
RPATH - I was just reading documentation for dlopen and I thought about the same thing because it would work for dynamic libraries that CoreRT will produce in the future.

sergiy-k replied:
I have just created a PR (https://github.com/dotnet/corert/pull/3592) to enable RPATH.

MichalStrehovsky commented on src/Common/src/TypeSystem/Ecma/EcmaMethod.cs:
```diff
@@ -428,7 +428,15 @@ public override PInvokeMetadata GetPInvokeMethodMetadata()
+            // check to see whether it is assigned on the containing assembly
+            if (dllImportSearchPath == DllImportSearchPath.None)
+            {
+                dllImportSearchPath = ((EcmaAssembly)Module).GetDllImportSearchPath();
```
I usually put an assert in places like this:

```csharp
Debug.Assert(Module is IAssemblyDesc, "Multi-module assemblies");
```

If we ever add support for multi-module assemblies, it's a nice breadcrumb to search for.

MichalStrehovsky commented on src/Common/src/TypeSystem/Interop/IL/MarshalHelpers.cs:
```diff
@@ -87,6 +87,12 @@ public static bool IsStructMarshallingRequired(TypeDesc typeDesc)
+        // other P/Invokes to use lazy resolution.
+        // In particular, these functions are necessary to get the current application
+        // path, so having them use lazy resolution would end up causing StackOverflow
+        private static readonly string[] s_lazyPInvokeDependencies = { "GetModuleFileNameW", "CoTaskMemAlloc", "CoTaskMemFree",
```
@luqunl Added a similar whitelist in MCG when he did a similar feature. I gave him code review feedback asking to come up with a scheme (custom attribute or something) that the class library can put on these DllImports so that MCG doesn't hardcode it.

I have no power to block MCG code reviews and don't care much about MCG hardcoding implementation details of the class library, but I do care about the CoreRT type system not to do such thing.

Please work with Luqun and Yi to come up with a way to annotate the DllImports.

jkotas replied:
I believe we need both:
- Allow specifying what to delay load vs. eager load/statically link on command line. It is important if you do not have a control over the library that is doing the PInvoke. Also, you may want to link the same library different ways in different projects.
- Allow marking individual PInvokes as eager load/static link. Mono has a prior art here: It treats __Internal as static link. I am not saying that it is a nice clean design, but we will need to support this hack if we ever want CoreRT to be reasonably compatible with Mono.

MichalStrehovsky commented on src/Common/src/TypeSystem/Interop/IL/MarshalHelpers.cs:
```diff
@@ -96,6 +102,16 @@ public static bool UseLazyResolution(MethodDesc method, string importModule, PIn
+            PInvokeMetadata metadata = method.GetPInvokeMethodMetadata();
+            for (int i = 0; i < s_lazyPInvokeDependencies.Length; i++)
+            {
+                if (s_lazyPInvokeDependencies[i] == metadata.Module)
```
If this actually ever evaluates to true, we might have a bug somewhere. Feels like this should be `metadata.Name`.

MichalStrehovsky commented on src/ILCompiler.Compiler/src/Compiler/Compilation.cs:
```diff
@@ -97,7 +97,7 @@ public ObjectNode GetFieldRvaData(FieldDesc field)
                 var pInvokeFixup = (PInvokeLazyFixupField)field;
                 PInvokeMetadata metadata = pInvokeFixup.PInvokeMetadata;
-                return NodeFactory.PInvokeMethodFixup(metadata.Module, metadata.Name);
+                return NodeFactory.PInvokeMethodFixup(metadata.Module, metadata.Name, metadata.DllImportSearchPath);
```
Might as well change this method to take `PInvokeMetadata` directly. This list is only going to get bigger (Does the method update LastError? Do we need to do A/W probing, etc.)

MichalStrehovsky commented on src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/PInvokeModuleFixupNode.cs:
```diff
@@ -45,6 +48,7 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
 
             builder.EmitZeroPointer();
             builder.EmitPointerReloc(nameSymbol);
+            builder.EmitInt((int)_dllImportSearchPath);
```
Will we need to add more flags in the future? If so, coming up with a more compact enum (that is a contract used for storage in the file only) that can be extended to represent all the flags might be a good idea. People tend to do strictly additive changes when adding new features and I don't want us to end up with dozens of wasted bits in the format. The `DllImportSearchPath` has holes in it.

shrah commented:
I already addressed the feedbacks by moving the BaseDirectoy to bootstrapper. However, as per our offline discussion I am closing this PR, because 1. @sergiy-k  is unblocked by the https://github.com/dotnet/corert/pull/3592 and 2. He has a duplicate PR doing the same AppContext. BaseDirectory for Unix thing through P/Invoke in https://github.com/dotnet/corert/pull/3588
/cc: @yizhang82
